runtime/build.gradle.kts

```diff
@@ -50,7 +60,19 @@ kotlin {
             }
         }
 
+        val commonJvmAndroid = create("commonJvmAndroid") {
```

Comment:
Common sourceSet that contains all code that used to be in the `jvm` sourceset, except the `Util` class

Reply:
According to https://kotlinlang.org/docs/mpp-share-on-platforms.html#configure-the-hierarchical-structure-manually:

> Kotlin doesn’t currently support sharing a source set for these combinations:
> * Several JVM targets
> * JVM + Android targets
> * Several JS targets

It seems that Gradle is happy building from the command line. But when I open a file from `commonJvmAndroid` in IntelliJ (e.g. `MessageExtensionsJvm.kt`), it fails to resolve the java-specific imports.

I'm not sure what to make of this. The IDE breakage is certainly annoying, but not a deal breaker if having the common source set makes our lives easier otherwise. But I'm worried that the Kotlin docs explicitly call this out as an unsupported combination. That makes me think there might be other gotchas we just haven't run into yet.

Reply:
An alternative approach here could be to create two modules: `:runtime` and `:runtime-common`.

We would then extract all the `Base64` stuff out of `:runtime-common` and this module would only only have the JVM compile target (not Android). This would also allow to get rid of the `:jvm-test-types` module

`:runtime` would then offer the Android and `Base64` stuff on top.  and `:runtime` would then consume `runtime-common` as a dependency.

I don't think this is especially easy to do (because `Base64` is tangled with the rest), but might be worth exploring.

As for the IDE autocomplete: that's a good point, we could also avoid this by duplicating all code across Android/JVM but that would obviously have it's own drawbacks.

How would you like to tackle this moving forward? I can try out the two module approach? Or would you defer this to a follow up ticket?

Reply:
I suspect the two module approach is going to be complex to implement, as you say. However, I just remembered about a workaround that we were using in pbandk with Kotlin 1.3 for shared source sets (before HPP support was added) that might work here too:

```kt
    jvm {
        val main by compilations.getting {
            defaultSourceSet {
                kotlin.srcDir("src/commonJvmAndroid/kotlin")
            }
    }
    android {
        val main by compilations.getting { // this probably needs to be adjusted to use Android's sourceset configuration syntax
            defaultSourceSet {
                kotlin.srcDir("src/commonJvmAndroid/kotlin")
            }
    }
```

In essence, not creating a separate sourceset for `commonJvmAndroid` but instead having both JVM and Android sourcesets reuse the source files from that directory. Can you see if that could work?

Reply:
This was a great suggestion! 🙏 

I tested in Android Studio and all imports are resolved + build works from command line. Could you please double check the same in Intellij?

Reply:
Seems to be working in IntelliJ too 👍